common/matrix/dense_kernels.hpp.inc

```diff
@@ -365,4 +365,76 @@ __global__ __launch_bounds__(default_block_size) void reduce_total_cols(
 }
 
 
-}  // namespace kernel
\ No newline at end of file
+template <size_type block_size, typename IndexType, typename ValueType>
+__global__ __launch_bounds__(block_size) void row_permute(
+    size_type num_rows, size_type num_cols,
+    const IndexType *__restrict__ perm_idxs, const ValueType *__restrict__ orig,
+    ValueType *__restrict__ result, size_type stride_orig)
+{
+    constexpr auto warps_per_block = block_size / config::warp_size;
+    const auto global_id =
+        thread::get_thread_id<config::warp_size, warps_per_block>();
+    const auto row_id = global_id / num_cols;
+    const auto col_id = global_id % num_cols;
+    if (row_id < num_rows) {
+        result[row_id * stride_orig + col_id] =
```

Comment:
Could we also allow different strides here? Otherwise we might need to make additional allocations in case the output doesn't have the correct stride. Or do we not want this?

Reply:
Sorry, you are right. I should be using the stride of the result here.